src/main/java/org/spongepowered/mod/text/selector/SpongeArgumentHolder.java

```diff
@@ -0,0 +1,152 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.mod.text.selector;
+
+import com.flowpowered.math.vector.Vector3d;
+import com.flowpowered.math.vector.Vector3f;
+import com.flowpowered.math.vector.Vector3i;
+import com.flowpowered.math.vector.Vector3l;
+import com.google.common.base.Function;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.SetMultimap;
+import com.google.common.collect.Sets;
+import org.spongepowered.api.text.selector.ArgumentHolder;
+import org.spongepowered.api.text.selector.ArgumentType;
+import org.spongepowered.api.util.annotation.NonnullByDefault;
+
+import java.util.List;
+import java.util.Set;
+
+@NonnullByDefault
+public class SpongeArgumentHolder<T extends ArgumentHolder<?>> implements ArgumentHolder<T> {
+
+    private final Set<T> subtypes;
+
+    @SuppressWarnings("unchecked")
+    protected SpongeArgumentHolder() {
+        if (this instanceof ArgumentType) {
+            this.subtypes = ImmutableSet.of((T) this);
+        } else {
+            throw new IllegalStateException("Must provide subtypes if not ArgumentType");
+        }
+    }
+
+    public SpongeArgumentHolder(Set<T> subtypes) {
+        this.subtypes = ImmutableSet.copyOf(subtypes);
+    }
+
+    @Override
+    public int getCount() {
+        return this.subtypes.size();
+    }
+
+    @Override
+    public Set<T> getTypes() {
+        return this.subtypes;
+    }
+
+    public static class SpongeVector3<V, T> extends SpongeArgumentHolder<ArgumentType<T>> implements ArgumentHolder.Vector3<V, T> {
+
+        private static final SetMultimap<Class<?>, Function<?, ?>> extractFunctionSets = HashMultimap.create();
+        static {
+            List<Class<?>> vectors = ImmutableList.<Class<?>>builder().add(Vector3i.class, Vector3l.class, Vector3f.class, Vector3d.class).build();
```

Comment:
Could be simplified a bit by doing
`Arrays.<Class<?>>asList(Vector3i.class, Vector3l.class, Vector3f.class, Vector3d.class);`